api/src/apps/routes/logs.rs

```diff
@@ -0,0 +1,374 @@
+use crate::{
+    apps::Apps,
+    http_result::HttpResult,
+    models::{AppName, AppNameError, LogChunk},
+};
+use chrono::DateTime;
+use futures::stream::StreamExt;
+use http_api_problem::HttpApiProblem;
+use reqwest::header::{ACCEPT, CONTENT_DISPOSITION, LINK};
+use rocket::{
+    http::{Accept, ContentType, RawStr, Status},
+    request::FromRequest,
+    response::stream::{Event, EventStream},
+    response::{Responder, Response},
+    Request, State,
+};
+use std::{str::FromStr, sync::Arc};
+
+#[get("/<app_name>/logs/<service_name>?<log_query..>", rank = 1)]
+pub(super) async fn logs<'r>(
+    _apt: AcceptingPlainText,
+    app_name: Result<AppName, AppNameError>,
+    service_name: &'r str,
+    log_query: LogQuery,
+    apps: &State<Arc<Apps>>,
+) -> HttpResult<LogsResponse<'r>> {
+    let app_name = app_name?;
+
+    let since = match log_query.since {
+        None => None,
+        Some(since) => match DateTime::parse_from_rfc3339(&since) {
+            Ok(since) => Some(since),
+            Err(err) => {
+                return Err(
+                    HttpApiProblem::with_title(http_api_problem::StatusCode::BAD_REQUEST)
+                        .detail(format!("{}", err))
+                        .into(),
+                );
+            }
+        },
+    };
+
+    let log_chunk = apps
+        .get_logs(&app_name, service_name, &since, &log_query.limit)
+        .await?;
+
+    Ok(LogsResponse {
+        log_chunk,
+        app_name,
+        service_name,
+        limit: log_query.limit,
+        as_attachment: log_query.as_attachment,
+    })
+}
+
+#[get(
+    "/<app_name>/logs/<service_name>?<log_query..>",
+    format = "text/event-stream",
+    rank = 2
+)]
+pub(super) async fn stream_logs<'r>(
+    app_name: Result<AppName, AppNameError>,
+    service_name: &'r str,
+    log_query: LogQuery,
+    apps: &'r State<Arc<Apps>>,
+) -> HttpResult<EventStream![Event + 'r]> {
+    let app_name = app_name.unwrap();
```

Comment:
To keep it uniform accross both functions - `let app_name = app_name?;`

Reply:
done, thanks for watching out.